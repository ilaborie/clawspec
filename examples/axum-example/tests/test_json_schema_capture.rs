#![allow(missing_docs)]

use anyhow::Context;
use clawspec_core::register_schemas;
use rstest::rstest;
use tracing::info;

use axum_example::observations::domain::{LngLat, PartialObservation};

mod common;
pub use self::common::*;

#[rstest]
#[tokio::test]
async fn test_json_schema_capture_without_explicit_registration(
    #[future] app: TestApp,
) -> anyhow::Result<()> {
    let mut app = app.await;

    register_schemas!(app, LngLat).await;

    info!("Testing JSON schema capture with only create observation endpoint");
    let new_observation = PartialObservation {
        name: "Test Bird for Schema Capture".to_string(),
        position: LngLat {
            lng: 12.4,
            lat: -25.1,
        },
        color: Some("blue".to_string()),
        notes: Some("Testing automatic schema capture".to_string()),
    };

    let created_id = app
        .post("/observations")?
        .json(&new_observation)?
        .await
        .context("should create observation")?
        .as_json::<u32>()
        .await
        .context("should deserialize created observation ID")?;

    info!("Created observation with ID: {}", created_id);

    // Generate OpenAPI spec to verify schemas were captured
    let openapi_spec = app.collected_openapi().await;

    // Write to a test-specific file to inspect
    insta::assert_yaml_snapshot!(openapi_spec);

    // Verify that the main schema is now captured automatically
    let components = openapi_spec.components.as_ref().expect("having components");
    let schemas = &components.schemas;

    // This should succeed now
    assert!(
        schemas.contains_key("PartialObservation"),
        "PartialObservation schema should be captured"
    );

    Ok(())
}
