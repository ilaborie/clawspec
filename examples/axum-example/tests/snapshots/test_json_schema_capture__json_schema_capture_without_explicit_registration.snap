---
source: examples/axum-example/tests/test_json_schema_capture.rs
expression: openapi_spec
---
openapi: 3.1.0
info:
  title: Bird Observation API
  description: "A comprehensive API for managing bird observations with support for multiple content types, file uploads, and bulk operations. This API demonstrates RESTful design patterns and provides comprehensive CRUD operations for bird observation data."
  contact:
    name: Bird Observation Team
    url: "https://birdwatch.example.com/support"
    email: api-support@birdwatch.example.com
  version: 1.0.0
servers:
  - url: "https://api.birdwatch.example.com/api"
    description: Production server
paths:
  /api/observations:
    post:
      tags:
        - observations
      description: Create observation
      operationId: post-observations
      parameters: []
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/PartialObservation"
            example:
              color: blue
              name: Test Bird for Schema Capture
              notes: Testing automatic schema capture
              position:
                lat: -25.1
                lng: 12.4
      responses:
        "201":
          description: Status code 201
          content:
            application/json:
              schema:
                type: integer
                format: int32
                minimum: 0
components:
  schemas:
    LngLat:
      type: object
      required:
        - lng
        - lat
      properties:
        lat:
          type: number
          format: double
        lng:
          type: number
          format: double
    PartialObservation:
      type: object
      required:
        - name
        - position
      properties:
        color:
          type:
            - string
            - "null"
        name:
          type: string
        notes:
          type:
            - string
            - "null"
        position:
          $ref: "#/components/schemas/LngLat"
tags:
  - name: observations
